neuron_receiver/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+
+            while left < right {
+                let mid = left + (right - left) / 2;
+                let mut temp_log = self.clone();
+                temp_log.entries.truncate(mid);
+                let temp_entries_json = serde_json::to_string(&temp_log).unwrap_or_default();
+
+                if temp_entries_json.len() <= max_body_size {
+                    entries_json = temp_entries_json;
+                    left = mid + 1;
+                } else {
+                    right = mid;
+                }
+            }
+        }
+        entries_json
+    }
+}
+
+// "ryjl3-tyaaa-aaaaa-aaaba-cai"
+pub const ICP_LEDGER_ID: Principal = Principal::from_slice(&[0, 0, 0, 0, 0, 0, 0, 2, 1, 1]);
+// "rrkah-fqaaa-aaaaa-aaaaq-cai"
+pub const NNS_GOVERNANCE_ID: Principal = Principal::from_slice(&[0, 0, 0, 0, 0, 0, 0, 1, 1, 1]);
+
+/// Call the list_neurons from the management canister.
+/// This canister needs to be a hot key of the neurons in order to have full neuron access.
+/// At most this endpoint can return 9_400 neurons following estimations.
+/// As the response size grows following s(n) = 212 * n + 359.
+pub async fn list_neurons(args: ListNeurons) -> Result<ListNeuronsResponse, String> {
+    let res_gov: Result<(ListNeuronsResponse,), (i32, String)> =
+        ic_cdk::api::call::call(NNS_GOVERNANCE_ID, "list_neurons", (args,))
+            .await
+            .map_err(|(code, msg)| (code as i32, msg));
+    match res_gov {
+        Ok((res,)) => Ok(res),
+        Err((code, msg)) => Err(format!(
+            "Error while calling Governance canister ({}): {:?}",
+            code, msg
+        )),
+    }
+}
+
+async fn manage_neuron(
+    command: Command,
+    neuron_id: NeuronId,
+) -> Result<ManageNeuronResponse, String> {
+    let arg = ManageNeuron {
+        id: None,
+        neuron_id_or_subaccount: Some(NeuronIdOrSubaccount::NeuronId(neuron_id)),
+        command: Some(command),
+    };
+
+    let res_gov: Result<(ManageNeuronResponse,), (i32, String)> =
+        ic_cdk::api::call::call(NNS_GOVERNANCE_ID, "manage_neuron", (arg,))
+            .await
+            .map_err(|(code, msg)| (code as i32, msg));
+    match res_gov {
+        Ok((res,)) => Ok(res),
+        Err((code, msg)) => Err(format!(
+            "Error while calling Governance canister ({}): {:?}",
+            code, msg
+        )),
+    }
+}
+
+#[derive(Debug)]
+pub enum DisburseError {
+    FailedToCall(String),
+    UnexpectedAnswer(ManageNeuronResponse),
+}
+
+impl From<String> for DisburseError {
+    fn from(e: std::string::String) -> Self {
+        Self::FailedToCall(e)
+    }
+}
+
+pub async fn disburse(
+    neuron_id: NeuronId,
+    to_account: Account,
+) -> Result<DisburseToNeuronResponse, DisburseError> {
+    let account_id: icp_ledger::AccountIdentifier = to_account.into();
+    let manage_neuron_response = manage_neuron(
+        Command::Disburse(Disburse {
+            amount: None,
+            to_account: Some(AccountIdentifier {
```

Comment:
How come the we need 2 different AccountIdentifiers? I get that the ic repo is boilerplate  but maybe we could simplify here by using the same type if possible